File: leetcode/1-two-sum/two-sum.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        

        // Solution 1 o(nlogn)
        // vector<pair<int,int>> pnums;
        // for(int i = 0; i < nums.size();i++){
        //     pnums.push_back({nums[i],i});
        // }

        // sort(pnums.begin(),pnums.end());

        // int i = 0 , j = pnums.size()-1;
        // while(i < j){
        //     int curr = pnums[i].first + pnums[j].first;
        //     if(curr == target){
        //         return {pnums[i].second,pnums[j].second};
        //     }else if(curr > target){
        //          j--;
        //     }else{
        //         i++;
        //     }
        // }
        // return {-1,-1}; // will never come here..

        // Solution 2

        unordered_map<int,int> ump;
        for(int i = 0 ; i < nums.size();i++){
             int curr = nums[i];
             if(ump.find(target-curr) != ump.end()){
                return {i,ump[target-curr]};
             }else{
                ump[curr] = i;
             }

        }
        return {-1,-1};


    }
};

File: leetcode/1-two-sum/two-sum.java
// Brute force => start for an index i and search for all j > i+1 that arr[i] + arr[j] == tar .. the timecomplexity is o(n2) which will fail as constraints are 10^4 .. so need to optimise .. and we can optimise the solution from o(n2) to o(nlog ) to o(n)
// o(nlogn) => use a pair array where firsr is value and second is index and sort it by value .. then use two pointer on sorted space ..

// o(n) == use map to store the values and when we get the complement return it ..
class Solution {
    public int[] twoSum(int[] nums, int target) {

        Map<Integer,Integer> mp = new HashMap<>();
        for(int i = 0 ; i < nums.length ;i++){
             int compliment = target - nums[i];
             if(mp.containsKey(compliment)){
                return new int[]{i,mp.get(compliment)};
             }
             mp.put(nums[i],i);
        }
        return new int[]{-1,-1};
    }
}